controllers/twophase/state_machine.go

```diff
@@ -0,0 +1,186 @@
+func (m *chaosStateMachine) run(ctx context.Context, targetPhase v1alpha1.ExperimentPhase) (bool, error) {
+	updated := false
+
+	duration, err := m.Chaos.GetDuration()
+	if err != nil {
+		m.Log.Error(err, "failed to get chaos duration")
+		return updated, err
+	}
+	if duration == nil {
+		zero := time.Duration(0)
+		duration = &zero
+	}
+	if m.Chaos.GetScheduler() == nil {
+		m.Log.Info("Scheduler should be defined currently")
+		return updated, fmt.Errorf("misdefined scheduler")
+	}
+
+	status := m.Chaos.GetStatus()
+	currentPhase := m.Chaos.GetStatus().Experiment.Phase
+	m.Log.Info("change phase", "current phase", currentPhase, "target phase", targetPhase)
+
+	// covers 6 conditions
+	if currentPhase == targetPhase {
+		return updated, nil
+	}
+
+	// covers 5 conditions
+	if currentPhase == v1alpha1.ExperimentPhaseFinished {
+		return updated, unexpectedInto(currentPhase, targetPhase)
+	}
+
+	// In this branch `targetPhase != v1alpha1.ExperimentPhaseRunning` is always true
+	// move from Running phase to any phase should recover the experiment.
+	// covers 5 conditions
+	if currentPhase == v1alpha1.ExperimentPhaseRunning {
+		updated = true
+
+		m.Log.Info("recovering", "current phase", currentPhase, "target phase", targetPhase)
+		err := m.Recover(ctx, m.Req, m.Chaos)
+		if err != nil {
+			status.FailedMessage = err.Error()
+
+			m.Log.Error(err, "fail to recover")
+			return updated, err
+		}
+		status.Experiment.Phase = targetPhase
+
+		now := time.Now()
+		status.Experiment.EndTime = &metav1.Time{
+			Time: now,
+		}
+
+		if status.Experiment.StartTime != nil {
+			status.Experiment.Duration = now.Sub(status.Experiment.StartTime.Time).String()
+		}
+
+		// Pause from a running phase should set the nextStart to now
+		// so that the Reconciler will start another cycle of running right
+		// after resume
+		if targetPhase == v1alpha1.ExperimentPhasePaused {
+			m.Chaos.SetNextStart(now)
+		}
+	}
+
+	// currentPhase could be ExperimentPhaseUnitialized or Waiting or Paused or Failed
+	// covers 4 * 5 = 20 conditions
+	if currentPhase != v1alpha1.ExperimentPhaseRunning {
+		if targetPhase == v1alpha1.ExperimentPhaseRunning {
+			updated = true
+
+			m.Log.Info("applying", "current phase", currentPhase, "target phase", targetPhase)
+			err := m.Apply(ctx, m.Req, m.Chaos)
+			if err != nil {
+				m.Log.Error(err, "fail to apply")
+
+				status.Experiment.Phase = v1alpha1.ExperimentPhaseFailed
+				status.FailedMessage = err.Error()
+
+				return updated, err
+			}
+			status.Experiment.Phase = targetPhase
+
+			startTime := time.Now()
+			nextStart, err := utils.NextTime(*m.Chaos.GetScheduler(), startTime)
+			if err != nil {
+				m.Log.Error(err, "failed to get the next start time")
+				return updated, err
+			}
+			nextRecover := startTime.Add(*duration)
+
+			counter := 0
+			// if the duration is too long, `nextRecover` could be after `nextStart`
+			// we can jump over a start to make sure `nextRecover` is before `nextStart`
```

Comment:
Hi @YangKeao I have another idea to achieve maybe an ideal situation, that for standard cron (* * * * *), we could follow the schedule in this cycle (which we missed the start time), rather than start at the next cycle (or the cycle after the next).

So there are two situation: we entered in **running** period, or **waiting** period. We could calculate the whole cron cycle duration, and get the `waitDuration` from `cycle - duration`, which would be helpful in the following.

1. Running period
When `time.Now().Before(nextStart - waitDuration)`, we could see we are in the running period. Entering state machine, still in this 20 conditions, but nextRecover should be `nextStart - waitTime`. And keep the same for others.

2. Waiting period
The same, when `time.Now().After(nextStart - waitDuration)`. This time we should separate a condition, from `Uninitialized` to `Waiting`, and the only thing to do is set next start. After that it could be the same with the `every` type cron.

Reply:
`cycle` is not a stable value, for example: `5/15 * * * *`, it means execute at each `<any hour>:05` and `<any hour>:15`.

Reply:
@STRRL Yes you're right I missed this point. Maybe we could implement a `Previous`, just like `Next` but get the previous start time from certain time, which would also need when implementing [pause for duration](https://github.com/chaos-mesh/chaos-mesh/pull/1016/files#diff-04c792dfed57630c1de983400da2e7639c6d5c3a929c0124f7b5c87aa9a19105R121-R124).
Actually someone creates a [PR about this feature](https://github.com/robfig/cron/pull/216) in cron repo, but the maintainer thinks it's outside the scope of that repo so we need to do this ourselves.

Reply:
> ~`cycle` is not a stable value, for example: `5/15 * * * *`, it means execute at each `<any hour>:05` and `<any hour>:15`.~
> 
> That's not an issue. I still considering other unexpected situations.

Yeah but something like `5/18 * * * *` would still cause an unstable cycle (run at 5, 23, 41, 59 in each hour)

Reply:
@Yiyiyimu I have come up with an idea: if the current phase is `Paused`, we should ignore the `targetPhase`. And loop and iterate forward the `nextStart` and `nextRecover` until at least one of them is bigger than `now`.

The behavior of `@every 5min` is different as it may not apply right after resume. However, it works like erasing a period of time.

Reply:
@YangKeao Yes it could work for `pause`, but that still could not solve the problem when we just start, since we don't know the last start time.

Reply:
@Yiyiyimu I have implemented the logic mentioned above. Could you have a look?